### PR TITLE
Add Python scanner and tests

### PR DESCRIPTION
**Pull request by grossag**
This change adds a python scanner that allows consumers to take dependencies
upon any packages imported by scripts that are run. It also adds tests to
verify proper functionality of the scanner.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: src/engine/SCons/Scanner/Python.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#
+
+__revision__ = "__FILE__ __REVISION__ __DATE__ __DEVELOPER__"
+
+import itertools
+import os
+import re
+import SCons.Scanner
+
+# Capture python "from a import b" and "import a" statements.
```
Comment (mwichmann) [inline src/engine/SCons/Scanner/Python.py:44]: how about `import foo as bar` and `from foo import bar as baz`?

Comment (grossag) [reply src/engine/SCons/Scanner/Python.py:44]: Good question. The first works but the second doesn't due to a bug in my regex. I will add test cases for both and update the diff.

File: src/engine/SCons/Scanner/PythonTests.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+import TestCmd
+
+import SCons.Node.FS
+import SCons.Scanner.Python
+
+test = TestCmd.TestCmd(workdir='')
+
+# This directory is a simple empty package.
+test.subdir('simple_package')
```
Comment (bdbaddog) [inline src/engine/SCons/Scanner/PythonTests.py]: Please switch this to a directory fixture? We're trying to phase out having the tests create each file needed every time it's run.  See: https://github.com/SCons/scons/wiki/TestingMethodology#directory-fixtures

Comment (grossag) [reply src/engine/SCons/Scanner/PythonTests.py]: Done, thanks.

**Comment by grossag** [comment]
I'm going to need someone to close the various Sider issues because it is scanning the test python scripts as though they are real scripts.

**Comment by bdbaddog** [comment]
> I'm going to need someone to close the various Sider issues because it is scanning the test python scripts as though they are real scripts.

Done.

**Comment by bdbaddog** [comment]
Looks like 4 python scanner tests are failing. Please take a look (at least in travis..)

File: testing/framework/test-framework.rst [hunk 3]
```diff
@@ -308,6 +308,9 @@ in that sense: "the fixture for this test is foo", instead of writing
 a whole bunch of strings to create files. Since these setups can be
 reusable across multiple tests, the *fixture* terminology applies well.
 
+Note: fixtures must not be treated by SCons as runnable tests. To exclude
+them, see instructions in the above section named "Finding Tests".
```
Comment (mwichmann) [inline testing/framework/test-framework.rst:312]: yup, this looks like a useful note.

**Comment by bdbaddog** [comment]
if a builder specifies a python file as source, will this scanner then include those dependencies on the target?

Might need some documentation changes?

**Comment by grossag** [comment]
> if a builder specifies a python file as source, will this scanner then include those dependencies on the target?
> 
> Might need some documentation changes?

I don't believe that it will have this effect. Specifically, if IMPLICIT_COMMAND_DEPENDENCIES is True, actions are indeed scanned but there is no recursion. The only code that I can find in SCons that does recursive dependency scanning is in Node.get_implicit_deps(), as long as the scanner in use defines recurse_nodes to actually do recursion. As a result, I don't believe that this will cause any changes in practice to what consumers see, but please correct me if I'm missing something.

I do have hopes of eventually introducing a new flag that opts into scanning Python files if they are found in implicit command dependencies, but this would not automatically happen.

**Comment by bdbaddog** [comment]
hmm in theory if the scanner's properly plumbed env.Command('myfile.py args") should walk the scanner for myfile.py and add the implicit dependencies.
that's the way the scanning system is supposed to work.. 

Are you calling:
SourceFileScanner.add_scanner('.xyz', XYZScanner)
(search for that in the manpage to see what it does..)

Likely what should be done is to add a Python tool which then adds this scanner to the global scanner list. (we'd default that off for now?)
Then there'd need to be some tests for that. 
If you can stuff strings into the scanner and check output, you might avoid file io all together in testing the scanner, and then have system test to add python tool and check that.

Sorry for the piecemealness of this, but sometimes as we get each bit done the next lightbulb goes one.

**Comment by bdbaddog** [comment]
I'd still like to see a Python Tool which would plug this scanner in to the appropriate logic. 
That said unless someone objects I don't see any reason to not merge as is.
Thoughts?

**Comment by grossag** [comment]
Coming back to this, @bdbaddog do you want me to do the Python tool as part of this pull request? If so, do you have thoughts on what this would look like? Last time around when you introduced it I wasn't able to understand what it would look like in practice.

The context I'm coming from is that the codebase I work on has a bunch of actions like "$PYTHON somescript.py $SOURCES" or "$PYTHON3 somescript.py $SOURCES", sometimes with various args included.

**Comment by grossag** [comment]
Any thoughts on what I should do next with this PR?

File: src/CHANGES.txt [hunk 4]
```diff
@@ -17,6 +18,9 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
     - Drop support for Python 2.7. SCons will be Python 3.5+ going forward.
 
 
+  From Adam Gross:
+    - Added new module SCons.Scanner.Python to allow scanning .py files.
```
Comment (bdbaddog) [inline src/CHANGES.txt:22]: Add something like "the python tool is not added by default. Please use env.Tool('python') to add this."

**Comment by bdbaddog** [comment]
Add a blurb on the manpage. There's a list of supported tools. Add 'python'

File: src/engine/SCons/Tool/python.xml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+<sconsdoc xmlns="http://www.scons.org/dbxsd/v1.0"
+          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://www.scons.org/dbxsd/v1.0 http://www.scons.org/dbxsd/v1.0/scons.xsd">
+
+<tool name="python">
+<summary>
+<para>
+Loads the Python scanner scanner into the invoking environment. When loaded, the scanner will
+attempt to find implicit dependencies for any Python source files in the list of sources
+provided to an actual that uses this environment.
```
Comment (mwichmann) [inline src/engine/SCons/Tool/python.xml:31]: Couldn't parse that: "to an actual" ?

File: src/engine/SCons/Tool/python.xml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+]>
+
+<sconsdoc xmlns="http://www.scons.org/dbxsd/v1.0"
+          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://www.scons.org/dbxsd/v1.0 http://www.scons.org/dbxsd/v1.0/scons.xsd">
+
+<tool name="python">
+<summary>
+<para>
+Loads the Python scanner scanner into the invoking environment. When loaded, the scanner will
```
Comment (mwichmann) [inline src/engine/SCons/Tool/python.xml:29]: waas "scanner scanner" intended?

Comment (grossag) [reply src/engine/SCons/Tool/python.xml:29]: oops no. @bdbaddog can you fix that on master?

Comment (grossag) [reply src/engine/SCons/Tool/python.xml:29]: Also "to an actual" should be "to an action". Sorry for the typos.

Comment (bdbaddog) [reply src/engine/SCons/Tool/python.xml:29]: will do.

**Pull request merged**